wasmbinding/stargate_whitelist_test.go

```diff
@@ -0,0 +1,114 @@
+}
+
+func TestStargateTestSuite(t *testing.T) {
+	suite.Run(t, new(StargateTestSuite))
+}
+
+func (suite *StargateTestSuite) TestDeterministicJsonMarshal() {
+	testCases := []struct {
+		name             string
+		originalResponse string
+		updatedResponse  string
+		queryPath        string
+		expectedProto    proto.Message
+	}{
+
+		/**
+		   * Origin Response
+		   * balances:<denom:"bar" amount:"30" > pagination:<next_key:"foo" >
+		   * "0a090a036261721202333012050a03666f6f"
+		   *
+		   * New Version Response
+		   * The binary built from the proto response with additional field address
+		   * balances:<denom:"bar" amount:"30" > pagination:<next_key:"foo" > address:"cosmos1j6j5tsquq2jlw2af7l3xekyaq7zg4l8jsufu78"
+		   * "0a090a036261721202333012050a03666f6f1a2d636f736d6f73316a366a357473717571326a6c77326166376c3378656b796171377a67346c386a737566753738"
+		   // Origin proto
+		   message QueryAllBalancesResponse {
+		  	// balances is the balances of all the coins.
+		  	repeated cosmos.base.v1beta1.Coin balances = 1
+		  	[(gogoproto.nullable) = false, (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"];
+		  	// pagination defines the pagination in the response.
+		  	cosmos.base.query.v1beta1.PageResponse pagination = 2;
+		  }
+		  // Updated proto
+		  message QueryAllBalancesResponse {
+		  	// balances is the balances of all the coins.
+		  	repeated cosmos.base.v1beta1.Coin balances = 1
+		  	[(gogoproto.nullable) = false, (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"];
+		  	// pagination defines the pagination in the response.
+		  	cosmos.base.query.v1beta1.PageResponse pagination = 2;
+		  	// address is the address to query all balances for.
+		  	string address = 3;
+		  }
+		*/
+		{
+			"Query All Balances",
+			"0a090a036261721202333012050a03666f6f",
+			"0a090a036261721202333012050a03666f6f1a2d636f736d6f73316a366a357473717571326a6c77326166376c3378656b796171377a67346c386a737566753738",
+			"/cosmos.bank.v1beta1.Query/AllBalances",
+			&banktypes.QueryAllBalancesResponse{
+				Balances: sdk.NewCoins(sdk.NewCoin("bar", sdk.NewInt(30))),
+				Pagination: &query.PageResponse{
+					NextKey: []byte("foo"),
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(fmt.Sprintf("Case %s", tc.name), func() {
+			suite.SetupTest()
+
+			binding, ok := wasmbinding.StargateWhitelist.Load(tc.queryPath)
+			suite.Require().True(ok)
+
+			originVersionBz, err := hex.DecodeString(tc.originalResponse)
+			suite.Require().NoError(err)
+			jsonMarshalledOriginalBz, err := wasmbinding.ConvertProtoToJsonMarshal(binding, originVersionBz, suite.app.AppCodec())
+			suite.Require().NoError(err)
+
+			newVersionBz, err := hex.DecodeString(tc.updatedResponse)
+			suite.Require().NoError(err)
+			jsonMarshalledUpdatedBz, err := wasmbinding.ConvertProtoToJsonMarshal(binding, newVersionBz, suite.app.AppCodec())
+			suite.Require().NoError(err)
```

Comment:
Do you think it makes sense to split "origin response" and "new version response" into 2 separate test cases?

I feel like we are testing several things in one test case

Reply:
sorry if I missed something, why / what do you think is getting tested in this single test case?
It seems to me that we're only testing if updated proto still returns same json marshalled bytes

Reply:
No probs at all, I could have been more clear. What I was thinking is to change:
```
originVersionBz, err := hex.DecodeString(tc.originalResponse)
suite.Require().NoError(err)
jsonMarshalledOriginalBz, err := wasmbinding.ConvertProtoToJsonMarshal(binding, originVersionBz, suite.app.AppCodec())
suite.Require().NoError(err)

newVersionBz, err := hex.DecodeString(tc.updatedResponse)
suite.Require().NoError(err)
jsonMarshalledUpdatedBz, err := wasmbinding.ConvertProtoToJsonMarshal(binding, newVersionBz, suite.app.AppCodec())
suite.Require().NoError(err)
```

to

```
decodedBuf, err := hex.DecodeString(tc.response)
suite.Require().NoError(err)
jsonMarshalledUpdatedBz, err := wasmbinding.ConvertProtoToJsonMarshal(binding, decodedBuf, suite.app.AppCodec())
suite.Require().NoError(err)
```

So that we have 2 separate test cases:
```
{
	"Query All Balances - Old Version",
	"0a090a036261721202333012050a03666f6f",
	"/cosmos.bank.v1beta1.Query/AllBalances",
	&banktypes.QueryAllBalancesResponse{
		Balances: sdk.NewCoins(sdk.NewCoin("bar", sdk.NewInt(30))),
		Pagination: &query.PageResponse{
			NextKey: []byte("foo"),
		},
	},
},
{
	"Query All Balances - New Version",
"0a090a036261721202333012050a03666f6f1a2d636f736d6f73316a366a357473717571326a6c77326166376c3378656b796171377a67346c386a737566753738",
	"/cosmos.bank.v1beta1.Query/AllBalances",
	&banktypes.QueryAllBalancesResponse{
		Balances: sdk.NewCoins(sdk.NewCoin("bar", sdk.NewInt(30))),
		Pagination: &query.PageResponse{
			NextKey: []byte("foo"),
		},
	},
},
```

Just to be clear, this is an optional comment. It just felt that we can split this one case into two to keep it more focused on the system under test (`wasmbinding.ConvertProtoToJsonMarshal`)

Reply:
Oh, I see what you mean. The main purpose of this test is not to `ConvertProtoToJsonMaral` but to test and ensure that stargate queries return the same, deterministic response, even though we have an update for the proto struct used in the state machine.

I added a separate test case for ConvertProtoToJsonMarshal while adding more comment on this test function to avoid future confusion